### PR TITLE
Add TGI documentation

### PR DESCRIPTION
**Pull request by aspctu**

### CODE REVIEW

File: docs/guides/tgi.mdx [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+- Galactica
+- GPT-Neox
+- Llama
+- OPT
+- SantaCoder
+- Starcoder
+- Falcon 7B
+- Falcon 40B
+- MPT
+- Llama V2
```
Comment (squidarth) [inline docs/guides/tgi.mdx:31]: nit, I'd roughly sort this in order of popularity

File: docs/guides/tgi.mdx [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+```
+- "truss predict": This CLI command invokes your model
+- "-d": This tells the CLI that the next argument is the data you want to pass to your model
+- '{"inputs": "What is a large language model?", "parameters": {"max_new_tokens": 128, "sample": true}}': This is the data you want to pass to your model. TGI expects a JSON object with 2 keys: `inputs` and `parameters` (optional). `inputs` is the prompt you want to pass to your model and `parameters` is a JSON object of parameters you want to pass to your model.
+- "--published": This tells the CLI that you want to invoke the production endpoint (vs. a development endpoint)
+```
+
+# Tuning your TGI server
+
+After deploying your model, you may notice that you're not getting the performance you'd like out of TGI. If you navigate to the `target_directory` from above, you'll find a `config.yaml` file that contains a key `build`. The following is a set of arguments you can pass to the `build` key to tune TGI for max performance:
```
Comment (squidarth) [inline docs/guides/tgi.mdx:107]: I would include an example here, something like:

```yaml config.yaml
model_name: ...
build:
    max_input_length: ...
```

File: docs/guides/tgi.mdx [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+1. `max_input_length`
```
Comment (squidarth) [inline docs/guides/tgi.mdx:109]: couple things, for each of these:
* include in this doc what the default is if there is one
* Phrase each of these as a "trade-off". ie: higher values of X mean this benefit and this drawback, while lower values mean this other benefit and this other drawback

File: docs/guides/tgi.mdx [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+This value is particularly important when maximizing GPU utilization.
+
+6. `max_waiting_tokens`
+
+This setting defines how many tokens can be passed before forcing the waiting queries to be put on the batch. Adjusting this value will help in optimizing the overall latency for end-users by managing how quickly waiting queries are allowed a slot in the running batch.
+
+Tune this parameter when optimizing for end-user latency and to manage the compute allocation between prefill and decode operations efficiently.
+
+7. `sharded`
```
Comment (squidarth) [inline docs/guides/tgi.mdx:138]: why would I set this to false ever?

File: docs/guides/tgi.mdx [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+Finally, you'll be asked for the name of your model.
+
+# Deploying your TGI model
+
+Now that we have a TGI model, let's deploy this model to Baseten and see how it performs.
+
+You'll need an API key to deploy your model. You can get one by navigating to your Baseten settings [page](https://app.baseten.co/settings/account/api_keys). To push the model to Baseten, run the following command:
+```
+(truss-py3.9) @aspctu $ truss push --publish
```
Comment (squidarth) [inline docs/guides/tgi.mdx:83]: get rid of everything before the "$" in these commands

File: docs/guides/tgi.mdx [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+Let's walk through what this command is doing
+```
+- "truss push": This CLI command zips and uploads your Truss to Baseten
+- "--publish": This tells Baseten that you want a Production endpoint (vs. a development environment)
+```
+
+After running this command, you'll be prompted to pass in your API key. Once you've done that, you can view your deployment status from the Baseten dashboard. Once the model has been deployed, you can invoke it by running the following command:
+
+```
+truss predict -d '{"inputs": "What is a large language model?", "parameters": {"max_new_tokens": 128, "sample": true}} --published'
```
Comment (squidarth) [inline docs/guides/tgi.mdx:95]: add "$" at begining of the line

File: docs/guides/tgi.mdx [hunk 7]
```diff
@@ -0,0 +1,140 @@
+---
+title: "High performance model serving with TGI"
+description: "A guide to using TGI for your model"
+---
+
+TGI, or Text Generation Inference, is a high performance model server for language models built by Huggingface. In this doc, we'll cover when to use TGI, how to tune TGI for performance and how to deploy it to Baseten.
```
Comment (squidarth) [inline docs/guides/tgi.mdx:6]: consider linking back to TGI on huggingface (https://huggingface.co/text-generation-inference) here

**Review by squidarth — APPROVED** [review]
awesome, thanks for writing this!

File: docs/guides/tgi.mdx [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+# What is TGI?
+
+TGI consists of 2 parts:
+
+1. A high performance, Rust-based server
+2. A set of optimized model implementations that outperform generic implementations
+
+To optimize GPU utilization and improve throughput / latency, it's common to implement optimizations at the server-level and the model-level.
```
Comment (philipkiely-baseten) [inline docs/guides/tgi.mdx:15]:
```suggestion
To optimize GPU utilization and improve throughput & latency, it's common to implement optimizations at the server level and the model level.
```

File: docs/guides/tgi.mdx (hunk 5 above)
Comment (philipkiely-baseten) [inline docs/guides/tgi.mdx:83]:
```suggestion
truss push --publish
```

File: docs/guides/tgi.mdx [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+```
+
+Let's walk through what this command is doing
+```
+- "truss init": This CLI command initializes a empty scaffold for a variety of model backends that truss supports (TGI, VLLM, Triton and TrussServer)
+- "./my-tgi-truss": The `init` command requires a `target_directory` which is where this empty scaffold will be generated
+- "--backend TGI": This tells truss that we are specifically interested in a TGI truss, which instantiates a different scaffold than the default (TrussServer)
+```
+
+After running that command, you'll be prompted to pass a `model_id`.
```
[the remark this answers is not in the document]
Comment (aspctu) [reply docs/guides/tgi.mdx:51]: unfortunately, this is the name of the parameter as TGI expects it.

**Pull request merged**